### PR TITLE
Move Processor Behavior to a Module

### PR DESCRIPTION
**Pull request by medcat**
I don't know if this is an acceptable pull request because of the lack 
of a contributing file, but I wanted to modify the way the processor
worked a little bit.  My primary concern with the processor was that
it was a class - you were forced to subclass the processor, which, in
my mind, made no sense.  The processor didn't make any sense as a 
class, mainly because of these reasons:

1. It violates the Liskov substitution principle, because the
 processor should never be instantiated within a program (it is 
 always subclassed). 
2. It's an anti-pattern.  Specifically, it's the 
[framework superclass anti-pattern][superclass-antipattern].
3. It violates the Composite resuse principle, or "composition over 
inheritance."

These are made complicated by the fact that the behaviors of the 
processor are essentially private.  the `#process` and `#process_all`
methods are not meant to be public, so describing how the processor
"behaves" publicly is awkward.

However, in interest of compatibility, I've kept the processor as a
class, and instead defined `AST::Processor::Module`, which contains
the previous processor behavior.  `AST::Processor` is now just a
class that includes `AST::Processor::Module`.

Anyway, nice library.  Cheers!

[superclass-antipattern]: http://michaelfeathers.typepad.com/michael_feathers_blog/2013/01/the-framework-superclass-anti-pattern.html

### CODE REVIEW

**Comment by medcat** [comment]
It looks like the build is failing for reasons independent to this change.

**Comment by whitequark** [comment]
Yes, you have a point. Two nits:

  * The docs should also move to `AST::Processor::Module`. Maybe also deprecate `AST::Processor` entirely, why not.
  * I don't like the name `Module`. That just describes how it's implemented, not what it does. Maybe `Mixin` or `Behavior`.

**Comment by medcat** [comment]
I agree with both points; how should I deprecate `AST::Processor`?

**Comment by whitequark** [comment]
`# @deprecated message...`

**Comment by medcat** [comment]
Done.

File: lib/ast/processor.rb [hunk 1]
```diff
@@ -1,266 +1,12 @@
 module AST
-  # Processor is a class which helps transforming one AST into another.
-  # In a nutshell, the {#process} method accepts a {Node} and dispatches
-  # it to a handler corresponding to its type, and returns a (possibly)
-  # updated variant of the node.
+  # This class includes {AST::Processor::Process}; however, it is no
+  # longer in use, since the module defines all of the behaviors that
```
Comment (whitequark) [inline lib/ast/processor.rb]: It is in use, of course, by all the current API consumers. It is deprecated however.

**Comment by whitequark** [comment]
`Process` is a really awful name. It's a verb, and a module is a thing.

**Comment by medcat** [comment]
I have to say, I'm extremely happy with the amount of documentation that you provide in your library.

**Comment by whitequark** [comment]
Renamed to `Mixin` instead and merged. Thanks!

**Comment by medcat** [comment]
What version can I expect to find this in?

**Comment by whitequark** [comment]
Released 2.1.0.

**Comment by medcat** [comment]
Awesome, thanks!

**Pull request merged**